add sparse guassian process module

SarahAlidoost opened the pull request:
closes #3

fnattino commented on src/fowt_ml/gaussian_process.py:
```diff
@@ -0,0 +1,289 @@
+        super().__init__(variational_strategy)
+
+        # covariance module: kernel: Prior information about latents
+        self.covar_module = gpytorch.kernels.ScaleKernel(
+            gpytorch.kernels.RBFKernel(batch_shape=torch.Size([num_latents])),
+            batch_shape=torch.Size([num_latents]),
+        )
+        # Mean module
+        self.mean_module = gpytorch.means.ConstantMean(
+            batch_shape=torch.Size([num_latents]),
+        )
```
I have noticed that `self.covar_module` and `self.mean_module` are basically functions (or at least behave as such within `forward()`?) - so maybe just call them `self.covar` and `self.mean`?

SarahAlidoost replied:
Right, first I thought the attributes might be used internally--defined in this [example](https://docs.gpytorch.ai/en/stable/examples/04_Variational_and_Approximate_GPs/SVGP_Multitask_GP_Regression.html#Types-of-Variational-Multitask-Models), but I renamed them as you suggested, and tests are still passing. I guess they are only used in the `forward` method.

fnattino commented on src/fowt_ml/gaussian_process.py:
```diff
@@ -0,0 +1,289 @@
+            batch_shape=torch.Size([num_latents]),
+        )
+
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
I see this appears in many places within different classes/functions here. Any reason why not to set the device globally (e.g. somewhere around line 20)?

fnattino commented on src/fowt_ml/gaussian_process.py:
```diff
@@ -0,0 +1,289 @@
+        # see check_multioutput_regressor
+        torch.set_default_dtype(torch.float64)
+
+    def __sklearn_tags__(self):
```
Didn't know about this method of sklearn estimator classes, nice!

fnattino commented on src/fowt_ml/gaussian_process.py:
```diff
@@ -0,0 +1,289 @@
+        y_train = _get_tensorlike(y_train)
+
+        # add some sklearn variables
+        self.X_ = x_train
+        self.y_ = y_train
```
Do sklearn estimators typically make a copy of the data they are trained on? Just trying to think why would this be needed.

SarahAlidoost replied:
yes, please see [Rolling your own estimator](https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator).

fnattino commented on src/fowt_ml/gaussian_process.py:
```diff
@@ -0,0 +1,289 @@
+        """Return the R^2 score of the prediction."""
+        y_pred = self.predict(x)
+        return r2_score(y, y_pred)
+
```
The `SparseGaussianModel` class below has a `calculate_score` method that does fitting and scoring at once, do we want the same here?

fnattino commented on src/fowt_ml/gaussian_process.py:
```diff
@@ -0,0 +1,289 @@
+        return scorer(self.estimator, x_test, y_test)
+
+
+def _get_tensorlike(array: ArrayLike | pd.DataFrame) -> torch.Tensor:
```
Since this function transforms an array- or dataframe-like to tensor, maybe calling it `_to_tensor()` makes more sense here?

SarahAlidoost replied:
Renamed!

fnattino reviewed (commented):
Hi @SarahAlidoost, nice! As my experience with ML and the torch-ecosystem is quite limited, I can not judge much on the model implementation. 

I have left a few comments/suggestions that are probably a little more general in nature - feel free to include them or not as you see fit 🙂

SarahAlidoost commented:
> Hi @SarahAlidoost, nice! As my experience with ML and the torch-ecosystem is quite limited, I can not judge much on the model implementation.
> 
> I have left a few comments/suggestions that are probably a little more general in nature - feel free to include them or not as you see fit 🙂

@fnattino thanks for reviewing. I addressed your comments. If the comments can be resolved now, can you please approve the PR? :pray:

fnattino reviewed (approved):
👍
